### ISSUE TITLE
Heuristic compilation database

### ISSUE THREAD

**Issue by sam-mccall** [clangd/clangd#349]
The need to create a compilation database to use clangd at all is a large hurdle.
For projects with complex builds we're probably stuck with this - running clang requires accurate flags. For projects with simple builds, we can fall back to guessing those flags. The MS VSCode plugin has some success with this approach.

Our current fallback is a simple function of the filename, which is pretty limited: we can't reasonably do a ton of IO in getFallbackCompileCommand(), and it can never support background indexing etc.
Creating a fallback *CDB* addresses these problems: it lets us amortize the discovery across files, gives us somewhere to put discovered state, and we can scan for files to index.

This should probably have several passes:
 - first we do the "regular" scan up from the source for compilation databases. We don't want to create a heuristic one if there's a real one further up.
 - then if we didn't find anything, we look for viable places to root a CDB. This should be fairly cheap (e.g. check for a few artifacts like makefiles) as we're looking at a bunch of candidates
 - if we select one, do detailed detection (determine include paths etc). This can do more work (e.g. recursive directory walking) but not an unbounded amount.

This is going to make invalidation more complicated. We don't do invalidation today, so this is "just" technical debt.

**Comment by sam-mccall** [comment]
Detection ideas:
 - version control root: .git etc
 - build systems we can't parse: Makefile, BUILD, CMakeLists (here but not in parent dir, for the root)
 - misc files that happen to live in project root a lot: configure.in, LICENSE.txt, package.json...
 - ad-hoc detection of particular projects: kernel, UE4, stdlib, who knows

**Comment by sam-mccall** [comment]
Misc stuff:
 - do we always want background-indexing on? for all files?
 - do we want all different types of detected projects to use a single set of heuristics with options, or should we split off e.g. UE4CDB and find ways to "mix in" the desired heuristic behaviors?
 - how much of this should we try to push into libtooling for other tools?

**Comment by sam-mccall** [comment]
Per-file ideas:
 - when the CDB is queried for a file, we can read its #include s and refine our guesses about where the -I lines should be rooted. We shouldn't solely rely on this though, cross-file information will make include-completion work.

**Comment by puremourning** [comment]
Gonna throw into the mix that when you find a file, asking the client to provide the flags would be _super great_ for YCM, as we can query the user's ycm_extra_conf file and for each discovered file, we can give you the correct flags (or None, if it's not a self-contained TU)

**Comment by sam-mccall** [comment]
> asking the client to provide the flags

Does this need to work for backrgound index? I'm worried about scale.
The critical case is when we open a big project that's already indexed. Before the index can be used, we need to a) create the CDB, b) list all the files, c) load all the shards and check if they're up to date (which depends on the compile command). Whether we do this during A or C, that's a lot of LSP traffic and editor-side processing can this be fast enough?

([`workspace/configuration`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_configuration) is likely the LSP vehicle for this)

**Comment by HighCommander4** [comment]
> Detection ideas:

Before we get to all the heuristic stuff mentioned, what about `InitializeParams.rootUri`?

>     * ad-hoc detection of particular projects: linux, UE4, stdlib, who knows

Wouldn't these fall more into the "For projects with complex builds we're probably stuck with [having them provide a CDB] - running clang requires accurate flags." case? Perhaps we could work with these projects to have their build systems include a "generate CDB" target?

**Comment by HighCommander4** [comment]
>   * maybe we can textually parse some useful bits from CMakeCache, build.ninja etc even if we can't understand them fully. Similarly ycm_extra_conf?!

If it's a CMake project, perhaps there's a way to "re-run CMake with the same arguments as last time", and we can just add the `-DCMAKE_EXPORT_COMPILE_COMMANDS=1`?

**Comment by puremourning** [comment]
> Does this need to work for backrgound index? 

yes ideally, as the user experience we all want is for the full index to just work. The use case here is where is have an _arbitrarily complex build system_ where I write some code (ycm_extra_conf) to query that build system for the right flags for a given TU. YCM has (had?) 1000s of users doing this sort of thing, and the number one request over the years is "why doesn't Go To Definition work". Of course, it _does_ work now that we use clangd but for users who can't make a compilation database, we'd love to offer the same or similar level of functionality.

> The critical case is when we open a big project that's already indexed. Before the index can be used, we need to a) create the CDB, b) list all the files, c) load all the shards and check if they're up to date (which depends on the compile command). Whether we do this during A or C, that's a lot of LSP traffic and editor-side processing can this be fast enough?

This is a good point. In the case of YCM (which is really ycmd) it's already a server process so this work would not be blocking the user. Perhaps in this mode, the requests can be batched or batched and throttled, but it's certainly worth thinking about.

**Comment by sam-mccall** [comment]
> Before we get to all the heuristic stuff mentioned, what about InitializeParams.rootUri?

I'm not sure this tells us anything meaningful about the codebase in practice. For `cd /tmp; $EDITOR ~/test.cc`:
 - vscode says null
 - coc.nvim says "/tmp"
 - ycm says $HOME

So maybe it tells you something about the editor! But similarly I'd guess coc is probably sensitive to your initial working directory, ycm to the first file you open, and vscode to the way you opened the window. I don't think it's reasonable for something so important to depend on this phase-of-the-moon stuff.

If the editor is *sure* it knows best, it can send the compile-commands directory, and we should respect that even if there's no "real" CDB and we need to use heuristics.

> If it's a CMake project, perhaps there's a way to "re-run CMake with the same arguments as last time", and we can just add the -DCMAKE_EXPORT_COMPILE_COMMANDS=1?

Yeah, for better or worse, CMake keeps all your settings from last time by default so that's `DCMAKE_EXPORT_COMPILE_COMMANDS=1`. It seems a bit rude to do this without asking first maybe, but we can prompt.

> Wouldn't these fall more into the "For projects with complex builds we're probably stuck with [having them provide a CDB]

Maybe! I meant more "there will be limits" than "we shouldn't try" though.
There's a reasonable concern that getting it half-working rather than obviously-broken will muddy the waters and discourage people from fixing their configs, leading to bad outcomes in the end. so we should be wary of UX and keep the user informed. We also shouldn't let the perfect be the enemy of the good in some of these cases though...

> it's already a server process so this work would not be blocking the user

It's not blocking keystrokes, but if it takes a minute instead of a few seconds to load the index then it might be blocking the user!
(We could separate out the validating the shards and put it on a different queue if it becomes a lot slower than loading them... everything is solvable with enough code)

**Comment by puremourning** [comment]
> I'm not sure this tells us anything meaningful about the codebase in practice. For cd /tmp; $EDITOR ~/test.cc:

> ycm says $HOME

I think that's a fairly unusual case. For 1000s of projects in the world with .ycm_extra_conf files, YCM would return the path of that file, which is most likely the project root. 

> It's not blocking keystrokes, but if it takes a minute instead of a few seconds to load the index then it might be blocking the user!

Blocking only if loading the background index is blocking real-time operations, which is isn't in the same way that _building_ the background index isn't blocking real-time operations in practice. I do take your point that for a large project, where the index is fully up-to-date, then the querying for the flags may dominate load time, but in any scenario where we have to parse the resulting TU, the expectation is that parsing time should dominate rather than dtermining-flags-time.

I do feel that the flexibility of providing arbitrary flags with arbitrary complexity for the very real projects where generating a compilation DB Is not something most users can do is worth the potential extra work.

To make this concrete, right now, in order to have a full index I have to run `bear make COMPILER=clang` and wait for the whole project to build (like 10 minutes) so that I have a compilation database. That's a lot more time than I would expect this process to take. 

I wonder if there's a middle-ground here. Like, could clangd even _record_ the compilation command used in the index. Then loading the index can be just-as-instant, and querying to see which TUs' compilation commands has changed could be done on a slow tick background process.

Just thinking out loud.

**Comment by rezamahdi** [comment]
I think you are making this so big.
take root directory of workspace. any .h/.hpp file is a header, any .cpp/.cc file is source.
run `c++`/`g++`/... command to find installed compiler and use this as driver to query system headers. (`--query-driver` flag)
